### DISCUSSION TITLE
Negative Jacobians During Thermal Expansion and Elastic Deformation

### DISCUSSION THREAD

**Discussion by wrkendrick** [idaholab/moose#28023] [Q&A Modules: Solid mechanics]
Hey all,

I'm running a pretty large mesh (generated via the Reactor module in MOOSE) doing a thermal solve w/ HeatConduction and deformation with IsotropicElasticityTensor, ComputeThermalExpansionEigenstrain, ComputeFiniteStrainElasticStress, and Physics/SolidMechanics/QuasiStatic. I'll provide an example of my use of each below.

The problem I'm running in to, is negative Jacobians specifically at certain TRI mesh blocks. The blocks where this happens are the TRIs in created from PolygonConcentricCircleMeshGenerator, almost all occurring in bottom of the geometry (where the reflector is defined). I've seen this happen before where a mesh element is small and the expansion inverts the mesh, but that's not what's happening here. All the volumes in the error report are positive, nothing stands out when I visually inspect the mesh, I even increased the size of the TRI elements and still get the errors.

Here's the geometry and the specific blocks that get the errors:

![badtris](https://github.com/idaholab/moose/assets/43858737/fe9267cf-c1e1-4532-9495-5c267c194d78)
![whole_geom](https://github.com/idaholab/moose/assets/43858737/1edc07f1-b95e-457e-81ed-0fac686dbea2)

Any ideas on what's causing this? Feel free to ask questions, I'm a bit in the dark of how to approach debugging this, so any help is very appreciated.

Edit: Also of note, my first attempt to debug this was setting the thermal expansion coefficient for the entire reflector material to 0. This worked for a handful of non-linear iterations, but eventually failed.

I'll attach my input and the error output so you can check it out. Apologies in advance for how large the input it, I generate it via a few scripts so it's not very human-readable.

[input.txt](https://github.com/user-attachments/files/16045409/input.txt)
[problems.txt](https://github.com/user-attachments/files/16045410/problems.txt)

Thanks y'all.

**Comment by GiudGiud** [comment]
Hello

I can see some really narrow tris in the problem.e mesh. Did you try to coarsen the mesh?
The mesh diagnostics generator detects thousands of non-planar sides.
```
*** Info ***
Number of non-planar element sides detected: 19388
```
Are these the ones on the outer part of cylinders? I think we expect those to be non planar on a second order mesh. That should be ok if so. We should still check that:
`n_azim * n_pins * n_axial` is equal to 19388

does this run with first order?

Guillaume

**Reply by wrkendrick** [reply]
Apologies, I uploaded incorrectly. Problem.txt is the exodus mesh of the problem subdomains reported from the error output, as you probably figured out. I ran a 2nd order run (slightly different geometry, same errors) and I'll attach the actual error output and the mesh generator file so you can check it out and rebuild the mesh yourself.

[core_gen.txt](https://github.com/user-attachments/files/16046594/core_gen.txt)
[error.txt](https://github.com/user-attachments/files/16046595/error.txt)

I'll check this number of sides.

I just ran a case with 1st order and it worked with no errors! Does that mean it's a 2nd order problem? Something to do with stress tensor calculations on the 2nd order tri elements?

I regenerated a new mesh using the `quad_center_elements = false` for all `PolygonConcentricCircleMeshGenerator`, so it may get further. I don't see any way to avoid TRI elements in `HexagonConcentricCircleAdaptiveBoundaryMeshGenerator` so it may still break.

**Reply by GiudGiud** [reply]
did you try with PRISM18?
To get those you have to extrude from a lower order TRI. So TRI6 instead of TRI7
see tri_element_type in 
https://mooseframework.inl.gov/source/meshgenerators/PolygonConcentricCircleMeshGenerator.html

EDIT I tried and the PeripheralRingMeshGenerator does not support higher order for now. i ll hit up the reactor module team

EDIT how did you even get PRISM21? I see PRISM18 from the conversion?

**Reply by GiudGiud** [reply]
There is two groups of problematic elements in error.txt.
(hmin = small side, hmax = largest side)
Some are 0.2 for hmin and 10 for hmax
and some are 0.03 for hmin and 10 for hmax
Both groups have a single side with 1 nullptr element? Do you know why that is? Are they on an actual external boundary of the mesh?

both are very thin. If we could have a less fine discretization (with less sectors) I think it could help
`quad_center_elements = false ` actually does not help make nice tri elements. With a quad element at the center, the mesh looks better imo

EDIT I tried and the PeripheralRingMeshGenerator does not support higher order for now. i ll hit up the reactor module team

**Reply by GiudGiud** [reply]
> Something to do with stress tensor calculations on the 2nd order tri elements?

that's possible. Or something to do with prism21 being finicky.
Note that you don't use all 21 nodes, in the error.txt output 3 nodes have no dofs

**Reply by GiudGiud** [reply]
Could you please share a version of the thermo-mechanics input that works with the mesh from core_gen.txt out of the box?
I am getting missing subdomains, I am guessing you used a different mesh for the input.i(txt) file
